luarules/gadgets/unit_xp_range_bonus.lua

```diff
@@ -38,17 +44,17 @@ function gadget:GameFrame(n)
             SetUnitMaxRange(unitID,newRange)
         end
     end
+
 	updateList = {}
 end
 
 function GG.requestMaverickExpUpdate(unitID)	
-	local unitDef = UnitDefs[GetUnitDefID(unitID)]
-	if(unitDef.customParams.rangexpscale == nil) then
+	if(XPDefs[attackerDefID] == nil) then
 		return
 	end
 
 	--schedule an update of range next frame when exp has been updated
-    updateList[unitID] = true
+    updateList[unitID] = UnitDefs[GetUnitDefID(unitID)]
```

Comment:
just pass the unitDefID instead of UnitDefs[unitDefID] you can get the unitDef later with XPDefs[unitDefID] only when you need it.

A small table lookup like `XPDefs[unitDefID]` is not expensive cpu-wise but storing the `unitDef` table for each and all units that range xp scale can be memory-wise (imagine making 1000 gunslingers)

Reply:
Solved.